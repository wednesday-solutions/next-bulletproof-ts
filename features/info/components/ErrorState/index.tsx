import { colors } from "@themes";
import { AlignCenter } from "common/styled";
import { FormattedMessage } from "react-intl";
import styled from "@emotion/styled";

const Error = styled.div`
  color: ${colors.error};
`;

const ErrorState = () => {
  return (
    <AlignCenter>
      <Error>
        <FormattedMessage id="something_went_wrong" />
      </Error>
    </AlignCenter>
  );
};

export default ErrorState;
