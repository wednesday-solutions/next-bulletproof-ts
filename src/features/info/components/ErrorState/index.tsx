import React from "react";
import { AlignCenter } from "@common/styled";
import { Trans } from "@lingui/macro";
import { styled } from "@mui/material/styles";

const ErrorComponent = styled("div")(({ theme }) => ({
  color: theme.palette.error.main,
}));

const ErrorState = () => {
  return (
    <AlignCenter>
      <ErrorComponent>
        <Trans>Something Went Wrong</Trans>
      </ErrorComponent>
    </AlignCenter>
  );
};

export default ErrorState;
