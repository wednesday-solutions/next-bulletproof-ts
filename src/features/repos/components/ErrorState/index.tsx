/**
 *
 * ErrorState
 *
 */

import React from "react";
import get from "lodash/get";
import { T, CustomCard } from "@common";
import { IResponse } from "@features/repos/api/getRecommendations";
import { Trans, t } from "@lingui/macro";

interface ErrorStateProps {
  loading: boolean;
  reposData: IResponse | undefined;
  reposError: string | undefined;
}

const ErrorState: React.FC<ErrorStateProps> = ({ reposData, reposError, loading }) => {
  let repoError: string | undefined;
  if (reposError) {
    repoError = reposError;
  } else if (!get(reposData, "totalCount", 0)) {
    repoError = t`Search Default`;
  }

  return !loading && repoError ? (
    <CustomCard color={reposError ? "red" : "grey"} data-testid="error-state">
      <T variant="subtitle2">
        <Trans>Repository List</Trans>
      </T>
      <T>{repoError}</T>
    </CustomCard>
  ) : null;
};

export default ErrorState;
