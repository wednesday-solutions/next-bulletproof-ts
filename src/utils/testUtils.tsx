import { messages } from "@app/translations/en";
import { i18n } from "@lingui/core";
import { I18nProvider } from "@lingui/react";
import { act, render, RenderOptions } from "@testing-library/react";
import React from "react";
import { Provider as ReduxProvider } from "react-redux";
import { SerializedStyles } from "@emotion/react";
import { store } from "../store";

/**
 * Renders the passed in components or tree with all the providers.
 * To configure just add more providers and use the function in tests to render with providers
 * @param children, all the components in the composition
 * @returns React.FC, renders the DOM with all the providers
 */
export const WithAllProviders: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  return (
    <ReduxProvider store={store}>
      <I18nProvider i18n={i18n}>{children}</I18nProvider>
    </ReduxProvider>
  );
};

/**
 * Generic render function that renders all the listed components with all the providers
 * @param ui The ReactElement (ie. Component) to render
 * @param options extra render options
 * @returns returns the RenderResult with all the things that react-testing library provides
 */
export const renderForTesting = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => {
  if (!i18n.locale) {
    // first client render
    i18n.load({ en: messages });
  }
  act(() => i18n.activate("en"));
  return render(ui, { wrapper: WithAllProviders, ...options });
};

/**
 * Stops execution in tests for a specified milliseconds
 * @param ms time in milliseconds, how long to stop
 * @returns returns a new Promise to hang
 */
export const timeout = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));

/**
 * Generate an API Response that can be used to mock API Responses
 * @param ok boolean specifing will the response resolve or reject
 * @param data the data that is to be sent back
 * @returns APIResponse consisting ok and data
 */
export function apiResponseGenerator<T>(ok: boolean, data: T) {
  return {
    ok,
    data,
  };
}

export const normalizeStyledCss = (styledCss: SerializedStyles) => {
  return styledCss.toString().trim().replace(/\s+/g, " ");
};

export * from "@testing-library/react";
export { renderForTesting as render };
