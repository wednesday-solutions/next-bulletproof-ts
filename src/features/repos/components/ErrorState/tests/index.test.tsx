/**
 *
 * Tests for ErrorState
 *
 */

import React from "react";
import { render } from "@utils/testUtils";
import ErrorState from "../index";

describe("<ErrorState />", () => {
  const props = {
    intl: {},
    loading: false,
    reposData: undefined,
    reposError: undefined,
  };

  it("should render and match the snapshot", () => {
    const { baseElement } = render(<ErrorState {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it("should contain 1 ErrorState component", () => {
    const { getAllByTestId } = render(<ErrorState {...props} />);
    expect(getAllByTestId("error-state").length).toBe(1);
  });
});
