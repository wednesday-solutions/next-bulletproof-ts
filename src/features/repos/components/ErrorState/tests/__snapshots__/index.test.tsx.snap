// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorState /> should render and match the snapshot 1`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.12);
  overflow: hidden;
}

.emotion-0.emotion-0 {
  margin: 20px 0;
  padding: 1rem;
  max-width: px;
  color: grey;
  color: grey;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.57;
  letter-spacing: 0.00714em;
}

.emotion-2 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
}

<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root emotion-0"
      color="grey"
      data-testid="error-state"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2 emotion-1"
        data-testid="t"
      >
        Repository List
      </h6>
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-2"
        data-testid="t"
      >
        Search Default
      </p>
    </div>
  </div>
</body>
`;
